File: DP_Inst/First_g.py
# n 個の品物(items に価値を格納) から任意の個数選ぶ
# 価値の総和が k を超えないうちで最高になるような組み合わせにおける、価値の総和を答える


n, k = map(int, input().split())
items = list(map(int, input().split()))

LIMIT = 100000

dp = [[False for _ in range(LIMIT + 1)] for _ in range(n + 1)]
dp[0][0] = True

for i in range(1, n + 1):
    for j in range(items[i - 1], LIMIT + 1):
        if not dp[i - 1][j - items[i - 1]]:
            continue
        # i 番目の品物を選ぶ場合
        dp[i][j] = True
        # 選ばない場合
        dp[i][j - items[i - 1]] = True

ans = max([(i if dp[n][i] else 0) for i in range(k + 1)])
print(ans)

File: DP_Inst/First_g.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;


int main(){
  // n 個の品物(items に価値を格納) から任意の個数選ぶ
  // 価値の総和が k を超えないうちで最高になるような組み合わせにおける、価値の総和を答える

  int n; cin >> n;
  int k; cin >> k;
  vector<int> items(n, 0);
  for(int i = 0; i < n; i++) cin >> items[i];

  constexpr int limit = 100000;

  vector<vector<bool>> dp(n + 1, vector<bool>(limit + 1, false));
  dp[0][0] = true;

  for(int i = 1; i <= n; i++){
    for(int j = items[i - 1]; j <= limit; j++){
      if(dp[i - 1][j - items[i - 1]]){
        // i 番目の品物を選ぶ場合
        dp[i][j] = true;
        // 選ばない場合
        dp[i][j - items[i - 1]] = true;
      }
    }
  }

  int ans = 0;
  for(int i = 0; i <= k; i++){
      if(dp[n][i]) ans = i;
  }
  cout << ans << endl;
}